Documentation/design-docs/host-tracing.md

```diff
@@ -0,0 +1,74 @@
+# Host tracing
+
+The various .NET Core host components provide detailed tracing of diagnostic information which can help solve issues around runtime, framework and assembly resolution and others.  
+
+## Existing support
+Currently (as of .NET Core 2.1) the host tracing is only written to the `stderr` output of the process. It can be turned on by setting `COREHOST_TRACE=1`.
+
+## Proposed changes
+For .NET Core 3 the proposal is to keep the existing behavior as described above and add two additional ways to capture the tracing:
+* Redirect the trace to a file instead - this would be done by setting `COREHOST_TRACE=1` and also `COREHOST_TRACEFILE=<path>` in the environment for the process.
+* For custom host support a way to capture the trace output in code. This will be done by adding `*_set_trace_listener` functions to `hostfxr` and `hostpolicy` which would let the custom host intercept all tracing.
+
+## Components involved
+Host tracing spans various components:
+* `dotnet` or `apphost` executable
+* `hostfxr` library
+* `hostpolicy` library
+
+If a custom host is used it will typically replace the executables and possible even some of the libraries. The custom host can use either `hostfx` or `hostpolicy` directly. So each of these components must support tracing on its own along with the configuration. In case several of these components are used together, they must cooperate and "share" the tracing setup.
+
+Currently (as of .NET Core 2.1) all these components include the same code for tracing but act completely independently. VIa the fact that they share the same code, the end behavior is that they all write their trace output to the same location `stderr`. The only synchronization currently implemented is that each component fully flushes all internal tracing buffers before transferring control to a different component.
+
+## New trace routing
+By default tracing is disabled and is not routed anywhere. If it's enabled it is routed through an instance of `host_trace_listener` interface.
+``` C++
+struct host_trace_listener
+{
+    void trace_verbose(const pal::char_t *message);
+    void trace_info(const pal::char_t *message);
+    void trace_warning(const pal::char_t *message);
+    void trace_error(const pal::char_t *message);
```

Comment:
Once we get into tracing assembly load events, there will be multiple threads sending trace events. The customer tracer will need to extract out the asynchronous trace messages. So, we should considering adding a ThreadID parameter to the callbacks.

Reply:
@steveharter What do you mean by "we might need more for ETW"? More methods or more parameters? Or both? Do you have an example in mind?

Reply:
@swaroop-sridhar That's a good point. I'm thinking we would add a second parameter to all the trace methods, something like `const pal::char_t *bindingId`. I plan to add something like a binding ID to the binding logging which will be added to the runtime as well. Thread ID would also work, but it's a bit confusing since there will be many bind events on the same thread. I think it would be better to generate unique ID for each bind event. We could pass it as a number, but that limits options for the future, so string feels a bit more future proof. What do you think?

Reply:
@vitek-karas having a bindingID parameter sounds good to me. I'm also OK having a string, but the only concern is that it may be less efficient (string comparison) for the listners to disentangle the messages from the various threads.